src/SparkConduit.sol

```diff
@@ -0,0 +1,314 @@
+import { WadRayMath } from 'aave-v3-core/contracts/protocol/libraries/math/WadRayMath.sol';
+
+import { UpgradeableProxied } from "upgradeable-proxy/UpgradeableProxied.sol";
+
+import { IInterestRateDataSource }          from './interfaces/IInterestRateDataSource.sol';
+import { ISparkConduit, IAllocatorConduit } from './interfaces/ISparkConduit.sol';
+
+interface PotLike {
+    function dsr() external view returns (uint256);
+}
+
+interface RolesLike {
+    function canCall(bytes32, address, address, bytes4) external view returns (bool);
+}
+
+interface RegistryLike {
+    function buffers(bytes32 ilk) external view returns (address buffer);
+}
+
+contract SparkConduit is UpgradeableProxied, ISparkConduit, IInterestRateDataSource {
+
+    using WadRayMath for uint256;
+
+    // Please note shares/pendingWithdrawals are in aToken "shares" instead of the underlying asset
+    struct Position {
+        uint256 shares;
+        uint256 pendingWithdrawals;
+    }
+
+    // Please note totalShares/totalPendingWithdrawals are in aToken "shares"
+    // instead of the underlying asset
+    struct AssetData {
+        bool enabled;
+        uint256 totalShares;
+        uint256 totalPendingWithdrawals;
+        mapping (bytes32 => Position) positions;
+    }
+
+    // -- Storage --
+
+    mapping(address => AssetData) private assets;
+
+    /// @inheritdoc ISparkConduit
+    address public roles;
+    /// @inheritdoc ISparkConduit
+    address public registry;
+    /// @inheritdoc ISparkConduit
+    uint256 public subsidySpread;
+
+    // -- Immutable/constant --
+
+    /// @inheritdoc ISparkConduit
+    IPool   public immutable pool;
+    /// @inheritdoc ISparkConduit
+    address public immutable pot;
+
+    uint256 private constant RAY = 10 ** 27;
+    uint256 private constant SECONDS_PER_YEAR = 365 days;
+
+    modifier auth() {
+        require(wards[msg.sender] == 1, "SparkConduit/not-authorized");
+        _;
+    }
+
+    modifier ilkAuth(bytes32 ilk) {
+        require(
+            RolesLike(roles).canCall(ilk, msg.sender, address(this), msg.sig),
+            "SparkConduit/ilk-not-authorized"
+        );
+        _;
+    }
+
+    constructor(IPool  _pool, address _pot) {
+        pool = _pool;
+        pot  = _pot;
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function deposit(bytes32 ilk, address asset, uint256 amount) external ilkAuth(ilk) {
+        require(assets[asset].enabled, "SparkConduit/asset-disabled");
+        require(
+            assets[asset].positions[ilk].pendingWithdrawals == 0,
+            "SparkConduit/no-deposit-with-pending-withdrawals"
+        );
+        require(amount <= maxDeposit(ilk, asset), "SparkConduit/max-deposit-exceeded");
+
+        address source = RegistryLike(registry).buffers(ilk);
+
+        require(
+            IERC20(asset).transferFrom(source, address(this), amount),
+            "SparkConduit/transfer-failed"
+        );
+
+        pool.supply(asset, amount, address(this), 0);
```

Comment:
Can't we get shares amount from return value of pool? Or are we intentionally ignoring and recalculating

Reply:
Lets use shares

Reply:
This doesn't return shares actually.